[MIG] 11.0 mig auth from http remote user

TDu opened the pull request.

lmignon committed:
[ADD] This module initialize the session by looking for the field HTTP_REMOTE_USER in the HEADER of the HTTP request and trying^Co bind the given value to a user

lmignon committed:
port auth_from_http_remote_user to 8.0

lmignon committed:
[PEP8] line lenght is now ridiculous (80 chars)

lmignon committed:
[FIX] if the default login page is disabled, redirect with *401*

lmignon committed:
Improve module description

lmignon committed:
Addons configuration parameter is now a field part of the Authentication group from base_setup[C

lmignon committed:
Improve doc

lmignon committed:
Code simplification: remove overkill method

lmignon committed:
Improve doc

lmignon committed:
[FIX] properlyt return the Unauthorized() response to avoid warning in the log file

lmignon committed:
[FIX] always check the uid in the session and the one in the request to avoid session mismatch...

lmignon committed:
[PEP8]

lmignon committed:
[FIX] return 'None' if no user name is found in the request hearder

lmignon committed:
If HTTP_REMOTE_USER is in the request headers and no corresponding user is found in odoo always issues Unauthorized (avoid redirect to the login page)
If the uid in the session is not the same as the one from the binded HTTP_REMOTE_USER, always logout to clean up the session

sbidoul committed:
auth_http_remote_user: slight reorganization and some comments (logic remains identical)

sbidoul committed:
auth_from_http_remote_user: remove unneeded browse

sbidoul committed:
auth_http_remote_user: test if already authenticated based on login instead of uid

Avoids a database query unless authentication is actually required.

lmignon committed:
[PEP-8]

pedrobaeza committed:
[IMP] auth_from_http_remote_user: README.rst

gurneyalex committed:
Add OCA as author of OCA addons

In order to get visibility on https://www.odoo.com/apps the OCA board has
decided to add the OCA as author of all the addons maintained as part of the
association.

gdgellatly committed:
Update Tests to Correctly Get DB

Fix introduced flake 8 errors

yvaucher committed:
Add missing default oca icons

gurneyalex committed:
remove deprecated test suite declarations

sbidoul committed:
[FIX] remove duplicate key in manifest

oca-transbot committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
[UPD] prefix versions with 8.0

oca-transbot committed:
OCA Transbot updated translations from Transifex

yajo committed:
Fix argument name that did not let users to sign up in runbot.

See https://github.com/OCA/server-tools/pull/277#discussion_r42502847

yajo committed:
Another typo.

yajo committed:
Do not recreate variable in case of dict().

oca-transbot committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
[FIX] remove en.po that was erroneously created by transbot

oca-transbot committed:
OCA Transbot updated translations from Transifex

sebalix committed:
[8.0] Fix Travis (#616)

* [FIX] auth_from_http_remote_user - Lint

* [FIX] server_environment - Lint

* [FIX] base_module_doc_rst - Lint

* [FIX] fetchmail_notify_error_to_sender - Fix XML view, it was the 'active' field from 'fetchmail_attach_from_folder' module which was targeted by mistake (belonging to another data model)

oca-transbot committed:
OCA Transbot updated translations from Transifex

TDu committed:
[MIG] auth_from_http_remote_user: Migration to 11.0

guewen commented on auth_from_http_remote_user/README.rst:
```diff
@@ -0,0 +1,174 @@
+To be active, the module must be installed in the expected databases and loaded
+at startup; Add the *--load* parameter to the startup command: ::
+
+  --load=web,web_kanban,auth_from_http_remote_user, ...
```
in v11 `web_kanban` is gone

TDu replied:
Removed

guewen commented on auth_from_http_remote_user/controllers/main.py:
```diff
@@ -0,0 +1,110 @@
+            request.session.authenticate(db_name, login=login,
+                                         password=key, uid=user_id.id)
+        except http.AuthenticationError as e:
+            raise e
```
The correct way to reraise an expection is only `raise` otherwise you lose where you are in the traceback.
In this case, I don't see why the handler is there.

TDu replied:
Thanks for the explanation, done.

guewen commented on auth_from_http_remote_user/controllers/main.py:
```diff
@@ -0,0 +1,110 @@
+        except Exception as e:
+            _logger.error("Error binding Http Remote User session",
+                          exc_info=True)
+            raise e
```
same note about reraising

TDu replied:
Done

guewen commented on auth_from_http_remote_user/models/res_users.py:
```diff
@@ -0,0 +1,76 @@
+    def copy(self, default=None):
+        default = default = {} if default is None else default
+        default['sso_key'] = False
+        return super(Users, self).copy(default)
```
The override of `copy` should be replaced by the `copy=False` argument in the definition of the `sso_key` field.

TDu replied:
Changed

guewen commented on auth_from_http_remote_user/models/res_users.py:
```diff
@@ -0,0 +1,76 @@
+                cls._uid_cache.setdefault(db, {})[uid] = passwd
+
+    @classmethod
+    def find_uid_from_iamuserid(cls, db='', login=''):
```
I don't think this one ought to be there?

TDu replied:
Oops, removed

guewen commented on auth_from_http_remote_user/controllers/main.py:
```diff
@@ -0,0 +1,110 @@
+                key = randomString(utils.KEY_LENGTH, '0123456789abcdef')
+                cr.execute('''update res_users
+                                set sso_key=%s
+                                where id=%s''', (key, user_id.id))
```
I guess there is a reason to bypass the ORM but could you put a comment explaining why?

TDu replied:
Done

guewen reviewed (changes requested):
Thanks! A few suggestions below

simahawk commented on auth_from_http_remote_user/README.rst:
```diff
@@ -0,0 +1,174 @@
+====================================
+
+This module initialize the session by looking for the field HTTP_REMOTE_USER in
+the HEADER of the HTTP request and trying to bind the given value to a user.
```
s/HEADER/HEADERS

TDu replied:
Changed

simahawk commented on auth_from_http_remote_user/README.rst:
```diff
@@ -0,0 +1,174 @@
+    :alt: Try me on Runbot
+    :target: https://runbot.odoo-community.org/runbot/149/11.0
+
+Known Issues / Roadmap
```
remove this section if empty

TDu replied:
Done

simahawk commented on auth_from_http_remote_user/__manifest__.py:
```diff
@@ -0,0 +1,36 @@
+##############################################################################
```
replace extended headers w/ short one

TDu replied:
Done

simahawk commented on auth_from_http_remote_user/__manifest__.py:
```diff
@@ -0,0 +1,36 @@
+    'category': 'Tools',
+    'author': "Acsone SA/NV,Odoo Community Association (OCA)",
+    'maintainer': 'ACSONE SA/NV',
+    'website': 'http://www.acsone.eu',
```
https://github.com/OCA/server-auth

TDu replied:
Done

simahawk commented on auth_from_http_remote_user/__manifest__.py:
```diff
@@ -0,0 +1,36 @@
+    'website': 'http://www.acsone.eu',
+    'depends': ['base', 'web', 'base_setup'],
+    "license": "AGPL-3",
+    'data': [],
```
remove empty keys

TDu replied:
Done

simahawk commented on auth_from_http_remote_user/models/res_users.py:
```diff
@@ -0,0 +1,76 @@
+        if not res:
+            try:
+                return super(Users, self).check_credentials(password)
+            except odoo.exceptions.AccessDenied:
```
same as noted above by Guewen: maybe I'm wrong but I don't see a good reason to catch the exception and raise it right after.

simahawk commented on auth_from_http_remote_user/tests/test_res_users.py:
```diff
@@ -0,0 +1,93 @@
+#
+##############################################################################
+
+from openerp.tests import common
```
s/openerp/odoo

TDu replied:
Changed this file and the others

simahawk commented on auth_from_http_remote_user/tests/test_res_users.py:
```diff
@@ -0,0 +1,93 @@
+
+    def test_login(self):
+        self.env.cr.execute('update res_users set sso_key = null;')
+        self.env.cr.commit()
```
is this commit really needed? can't we use a write on res.users. Also, shouldn't the key be empty already at this stage?

TDu replied:
Not sure but are we not bypassing the ORM because the authentication methods uses a different env ?

simahawk replied:
that's not my concern :) My concern is that if you commit you are going to commit any records that might be in the pool and when you reuse a test db (like when you are in dev mode) you can have strange behaviors/errors because of persistent test records.

TDu replied:
Well tried removing it and the test failed.

guewen replied:
Isn't because of the cache? Could you try `self.env.clear()`?

TDu replied:
Nope tests failed as well with `self.env.clear()`
Maybe because the authentication process by pass the cache ?

simahawk commented on auth_from_http_remote_user/tests/test_res_users.py:
```diff
@@ -0,0 +1,93 @@
+        self.env.cr.execute('update res_users set sso_key = null;')
+        self.env.cr.commit()
+        #
+        res_users_obj = self.registry('res.users')
```
self.env

TDu replied:
Done

simahawk commented on auth_from_http_remote_user/tests/test_res_users.py:
```diff
@@ -0,0 +1,93 @@
+        res = res_users_obj.authenticate(
+            common.get_db_name(), 'admin', token, None)
+        self.assertFalse(res)
+        # mimic what the new controller do when it find a value in
```
s/find/finds

TDu replied:
done

simahawk commented on auth_from_http_remote_user/tests/test_res_users.py:
```diff
@@ -0,0 +1,93 @@
+            common.get_db_name(), 'admin', token, None)
+        self.assertFalse(res)
+        # mimic what the new controller do when it find a value in
+        # the http header (HTTP_REMODE_USER)
```
s/REMODE/REMOTE

TDu replied:
done

simahawk commented on auth_from_http_remote_user/tests/test_res_users.py:
```diff
@@ -0,0 +1,93 @@
+                     ' of the mail module has created the column '
+                     '`notification_email_send` as REQUIRED into the table '
+                     'res_partner. BTW, it\'s no more possible to copy a '
+                     'res_user without an intefirty error')
```
s/intefirty/integrity

simahawk replied:
just to check: is this still needed?

TDu committed:
Remove unnecessary overriding of check method

The parent method calls check_credentials which is also implemented here,
so this is not needed.

TDu committed:
Remove unnecessary test if module installed or not

TDu committed:
Remove direct sql queries in favor of ORM

simahawk commented on auth_from_http_remote_user/tests/test_res_users.py:
```diff
@@ -24,12 +26,25 @@ def mock_cursor(cr):
+        reg = registry(self.env.cr.dbname)
+        with api.Environment.manage():
+            with reg.cursor() as cr:
+                env = api.Environment(cr, 1, {})
```
you are already admin here, you can use self._uid or self.env.user.id

simahawk commented on auth_from_http_remote_user/README.rst:
```diff
@@ -0,0 +1,171 @@
+are prompted for a login and password outside Odoo and are automatically
+logged in the system.
+
+.. [#] Shibolleth, Tivoli access manager, ..
```
*Shibboleth

simahawk commented on auth_from_http_remote_user/README.rst:
```diff
@@ -0,0 +1,171 @@
+logged in the system.
+
+.. [#] Shibolleth, Tivoli access manager, ..
+.. [#] Based on a ubuntu 12.04 env
```
still 12.04? or 16.04 now?

TDu replied:
Well this doc is based and tested on 12.04 (I guess) and not by me. So would it not be misleading to change the version number ?

simahawk commented on auth_from_http_remote_user/controllers/main.py:
```diff
@@ -0,0 +1,62 @@
+            else:
+                request.session.logout(keep_db=True)
+        try:
+            user = utils.get_user(request.env['res.users'], login)
```
any benefit in moving this to an util? Why not an instance method and `self.get_user`? It's easier to override it if needed too.

TDu replied:
I just did not think it belongs here and it would be needed to be overriden.
But if you think it is a bad idea I will change it

simahawk commented on auth_from_http_remote_user/controllers/main.py:
```diff
@@ -0,0 +1,62 @@
+            # later on is done in a specific environment as well
+            with api.Environment.manage():
+                with request.env.registry.cursor() as cr:
+                    env = api.Environment(cr, 1, {})
```
user odoo.SUPERUSER_ID

simahawk commented on auth_from_http_remote_user/models/res_users.py:
```diff
@@ -0,0 +1,34 @@
+
+    def check_credentials(self, password):
+        """Check credentials for SSO user"""
+        res = self.sudo().search([('id', '=', self._uid),
```
self.env.user.id

TDu replied:
This how it is done in the overridden method
https://github.com/odoo/odoo/blob/11dff2b05c72adb01f4094aa5077218a29315dec/odoo/addons/base/res/res_users.py#L441

simahawk commented on auth_from_http_remote_user/tests/test_res_users.py:
```diff
@@ -0,0 +1,92 @@
+                     ' installed. Even if `auth_from_http_remote_user` does '
+                     'not require the `mail` module, The previous installation'
+                     ' of the mail module has created the column '
+                     '`notification_email_send` as REQUIRED into the table '
```
just asking: is this still true? I don't see this field on v11

simahawk commented on auth_from_http_remote_user/utils.py:
```diff
@@ -0,0 +1,27 @@
+    user = users.sudo().search([
+        ('login', '=', login),
+        ('active', '=', True)
+        ])
```
limit=1

TDu commented:
@guewen @simahawk This one is in need of another look. I do think it is ready.

simahawk commented on auth_from_http_remote_user/__manifest__.py:
```diff
@@ -0,0 +1,17 @@
+    'website': 'https://github.com/OCA/server-auth',
+    'depends': ['base', 'web', 'base_setup'],
+    "license": "AGPL-3",
+    "installable": True,
```
you can drop these flags

TDu replied:
All three last ones  (installable, auto_install and application) ?

TDu replied:
Well that is what I did. I am curious where I can find the default values for those in the code ?

TDu replied:
Default values are here https://github.com/odoo/odoo/blob/0050019390daa759ac5882ded79a6ee50ab4749e/odoo/modules/module.py#L310
So indeed they can all go.

simahawk commented on auth_from_http_remote_user/controllers/main.py:
```diff
@@ -0,0 +1,83 @@
+            return user[0]
+        return None
+
+    def logging_http_remote_user(self, env, user):
```
logging or login? not clear IMO :)

TDu replied:
Indeed 'logging' is more about cutting trees, which I want nothing to do with :smile: 
Changed, thanks for your review.

simahawk reviewed (changes requested):
LGTM, minor comments

simahawk commented:
@lmignon would you mind passing by? :)

TDu committed:
fixup! [MIG] auth_from_http_remote_user: Migration to 11.0

TDu committed:
Add some more tests
